source/common/microkernel_architecture/microkernel_modules/instructions/AInstruction.hpp

```diff
@@ -0,0 +1,121 @@
+#ifndef AINSTRUCTION_HPP_
+# define AINSTRUCTION_HPP_
+
+# include "AInstructionBusClient.hpp"
+# include "struct_packed.h"
+# include <QByteArray>
+
+typedef unsigned int magicNumber_t;
+
+class AInstruction
+{
+protected:
+    //
+    // `struct instructionHeader_t` represents how the data is
+    // supposed to be organized inside `QByteArray _data`
+    // It should represent the first bytes of `_data`
+    //
+    // These bytes will be followed by the data representing the
+    // parameters, if `numberOfParameters` > 0
+    // Note: The parameters data representation is described below
+    //
+    STRUCT_PACKED(instructionHeader_t,
+        magicNumber_t magicNumber;
+        AInstructionBusClient::eClientId firstTransmitterId;
+        AInstructionBusClient::eClientId finalReceiverId;
+        unsigned int instructionCode;
+        unsigned int returnType;
+        int numberOfParameters;
+    );
+    struct instructionHeader_t;
+
+    //
+    // `struct instructionParameterHeader_t` represents how the data
+    // is supposed to be organized inside `QByteArray _data`, after
+    // the bytes representing the instruction header
+    //
+    // The parameter header will be followed by
+    // `parameterSize` bytes, these are the parameter data
+    //
+    // The next parameter, if any, is directly after the last
+    // parameter byte, symbolized by `// parameterData[parameterSize]`
+    //
+    STRUCT_PACKED(instructionParameterHeader_t,
+        unsigned int parameterType;
+        int parameterSize;
+        // byte_t parameterData[parameterSize];
+    );
+    struct instructionParameterHeader_t;
+
+    //
+    // The following values are often used during the creation
+    // or during the reading/filling of an `AInstruction`
+
+    static const magicNumber_t _MAGIC_NUMBER;
+
+    //
+    // Regarding `instructionHeader_t`
+    static const int _INSTRUCTION_HEADER_SIZE;
+    static const int _FIRST_PARAMETER_OFFSET;
+    //
+    // Regarding `instructionParameterHeader_t`
+    static const int _INSTRUCTION_PARAMETER_HEADER_SIZE;
+    static const int _PARAMETER_DATA_OFFSET;
+
+public:
+    class Parameter;
```

Comment:
Look at this, it's a nested class
https://github.com/EIP-SAM/SAM-Solution/blob/microkernel_arch_setup/source/common/microkernel_architecture/microkernel_modules/instructions/AInstructionParameter.hpp